invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,312 @@
+        self.post_migrate.append(callback)
+
+
+class MigrationSet:
+    """A set of Migrations. Performs validation during migration registration and provides utility methods."""
+
+    def __init__(self) -> None:
+        self._migrations: set[Migration] = set()
+
+    def register(self, migration: Migration) -> None:
+        """Registers a migration."""
+        if any(m.from_version == migration.from_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration from {migration.from_version} already registered")
+        if any(m.to_version == migration.to_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration to {migration.to_version} already registered")
+        self._migrations.add(migration)
+
+    def get(self, from_version: int) -> Optional[Migration]:
+        """Gets the migration that may be run on the given database version."""
+        # register() ensures that there is only one migration with a given from_version, so this is safe.
+        return next((m for m in self._migrations if m.from_version == from_version), None)
+
+    def validate_migration_path(self) -> None:
+        """
+        Validates that the migrations form a single path of migrations from version 0 to the latest version.
+        Raises a MigrationError if there is a problem.
+        """
+        if self.count == 0:
+            return
+        if self.latest_version == 0:
+            return
+        current_version = 0
+        touched_count = 0
+        while current_version < self.latest_version:
+            migration = self.get(current_version)
+            if migration is None:
+                raise MigrationError(f"Missing migration from {current_version}")
+            current_version = migration.to_version
+            touched_count += 1
+        if current_version != self.latest_version:
+            raise MigrationError(f"Missing migration to {self.latest_version}")
+        if touched_count != self.count:
+            raise MigrationError("Migration path is not contiguous")
+
+    @property
+    def count(self) -> int:
+        """The count of registered migrations."""
+        return len(self._migrations)
+
+    @property
+    def latest_version(self) -> int:
+        """Gets latest to_version among registered migrations. Returns 0 if there are no migrations registered."""
+        if self.count == 0:
+            return 0
+        return sorted(self._migrations, key=lambda m: m.to_version)[-1].to_version
+
+
+def get_temp_db_path(original_db_path: Path) -> Path:
+    """Gets the path to the temp database."""
+    return original_db_path.parent / original_db_path.name.replace(".db", ".db.temp")
+
+
+def get_backup_db_path(original_db_path: Path) -> Path:
+    """Gets the path to the final backup database."""
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+    return original_db_path.parent / f"{original_db_path.stem}_backup_{timestamp}.db"
```

Comment:
This is where the original database is backed up to before we move the migrated copy back into its place